### PR TITLE
Task 1285177: [LSP] Support Pull Diagnostics correctly for local client

### PR DESCRIPTION
**Pull request by LinglingTong**
Changes in this PR:

1. Add HintedSuggestion to XamlDiagnosticSeverity and handle this new severity in AbstractPullDiagnosticHandler.

What is HintedSuggestion?
HintedSuggestion is a new error Severity XAML added lately to support errors that show the "..." marker in editor but don't show up in the error list.

2. Switch to listening to Workspace.DocumentClosed event instead of RDT OnBeforeLastDocumentUnlock in XamlProjectService.  RunningDocumentTableEventTracker is also listening to OnBeforeLastDocumentUnlock and will fire a DocumentClosed event for it. This OnBeforeLastDocumentUnlock inside RunningDocumentTableEventTracker is doing the same check as we have in XamlProjectService. There is no need for us to do duplicate work and meanwhile as we remove the XAML file when the file gets closed this will ensure the file only gets removed after the DocumentClosed event gets fired.

### CODE REVIEW

**Comment by LinglingTong** [comment]
@mgoertz-msft @CyrusNajmabadi

File: src/VisualStudio/Xaml/Impl/Implementation/XamlProjectService.cs [hunk 1]
```diff
@@ -232,8 +239,13 @@ private void OnDocumentMonikerChanged(uint docCookie, IVsHierarchy hierarchy, st
             }
         }
 
-        private void OnDocumentClosed(string filePath)
+        private void OnDocumentClosed(string? filePath)
         {
```
Comment (mgoertz-msft) [inline src/VisualStudio/Xaml/Impl/Implementation/XamlProjectService.cs:243]: Is anyone else calling this? If not then we might as well put it in the event handler itself.

File: src/VisualStudio/Xaml/Impl/Implementation/XamlProjectService_IVsRunningDocTableEvents3.cs [hunk 2]
```diff
@@ -4,7 +4,9 @@
 
 #nullable disable
 
+using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
 using Microsoft.VisualStudio.Shell.Interop;
```
Comment (mgoertz-msft) [inline src/VisualStudio/Xaml/Impl/Implementation/XamlProjectService_IVsRunningDocTableEvents3.cs:8]: Why does removing code require new usings?

Comment (LinglingTong) [reply src/VisualStudio/Xaml/Impl/Implementation/XamlProjectService_IVsRunningDocTableEvents3.cs:8]: Oops it shouldn't. Don't remember how did they get added. I will clean it up.

---
In reply to: [606385567](https://github.com/dotnet/roslyn/pull/52369#discussion_r606385567) [](ancestors = 606385567)

File: src/VisualStudio/Xaml/Impl/Implementation/XamlProjectService.cs (hunk 1 above)
Comment (LinglingTong) [reply src/VisualStudio/Xaml/Impl/Implementation/XamlProjectService.cs:243]: Will do.

---
In reply to: [606385139](https://github.com/dotnet/roslyn/pull/52369#discussion_r606385139) [](ancestors = 606385139)

Comment (Youssef1313) [reply src/VisualStudio/Xaml/Impl/Implementation/XamlProjectService.cs:243]: The method is actually called in the other part of this partial type.

Comment (LinglingTong) [reply src/VisualStudio/Xaml/Impl/Implementation/XamlProjectService.cs:243]: If you mean the one inside OnBeforeLastDocumentUnlock, we have removed the call in this PR too :)

---
In reply to: [606391355](https://github.com/dotnet/roslyn/pull/52369#discussion_r606391355) [](ancestors = 606391355)

Comment (Youssef1313) [reply src/VisualStudio/Xaml/Impl/Implementation/XamlProjectService.cs:243]: @LinglingTong Ah, can indeed be put in the event handler itself then

**Review by mgoertz-msft — APPROVED** [review]
:shipit:

**Review by CyrusNajmabadi — APPROVED** [review]

**Pull request merged**